### PR TITLE
Site Switcher: Improve transition state when clicking on 'Add new site'

### PR DESCRIPTION
**Pull request by danielbachhuber**
See #67497 and conversation in p4TIVU-aiA-p2

Builds off #68373

## Proposed Changes

* Sets `state.ui.signupIsInitializing=true` at the beginning of the signup controller execution, and `state.ui.signupIsInitializing=false` at the end.
* Nullifies the primary content when `signupIsInitializing=true`, to prevent faulty intermediate state from rendering.
* Attempts to prevent the site switcher from closing when `signupIsInitializing=true`. This doesn't quite work though.

## Testing Instructions

TBD

### CODE REVIEW

File: client/layout/index.jsx [hunk 1]
```diff
@@ -300,7 +303,7 @@ class Layout extends Component {
 						{ this.props.secondary }
 					</div>
 					<div id="primary" className="layout__primary">
-						{ this.props.primary }
+						{ this.props.signupIsInitializing ? '' : this.props.primary }
```
Comment (danielbachhuber) [inline client/layout/index.jsx:306]: Are there any negative externalities from causing a full re-render here?

File: client/layout/style.scss [hunk 2]
```diff
@@ -18,6 +18,10 @@
 	transition: transform 0.15s ease-in-out, opacity 0.15s ease-out;
 }
 
+.is-signup-initializing .layout__secondary .site-selector {
+	transition-delay: 10s;
```
Comment (danielbachhuber) [inline client/layout/style.scss:22]: This doesn't do what I want it to

File: client/layout/index.jsx (hunk 1 above)
Comment (p-jackson) [reply client/layout/index.jsx:306]: Not due to re-rendering. Rendering shouldn't have any side effects so re-rendering is always safe.

But it's possible that mounting could cause issues. This change means that the `componentDidMount` methods and `useEffect` callbacks won't be called until after `indicateEndSignupStateChanges`. I guess theoretically it's possible that some of this code depends on things like the site ID being `null` at the time of mount.
I don't think it'll cause issues, but it is possible ...

File: client/layout/style.scss (hunk 2 above)
Comment (p-jackson) [reply client/layout/style.scss:22]: It must be because a different one of those other selectors is being used. I used this scss trick instead and it worked.

```scss
.layout__primary .main,
.layout__secondary,
.layout__secondary .sidebar,
.layout__secondary .site-selector {
	transition: transform 0.15s ease-in-out, opacity 0.15s ease-out;

	// We don't want to see the sidebar transition when navigating to the signup framework
	.is-signup-initializing & {
		transition-delay: 10s;
	}
}
```

It sometimes renders a visual glitch though. It doesn't happen every time, but sometimes after a second a second sidebar quickly appears. It doesn't do it every time.

https://user-images.githubusercontent.com/1500769/193479780-fe1d3a30-685b-45ce-a3e4-f3b28a809b61.mov

What if we just hide `secondary` when `signupIsInitializing` too? So it's a completely blank screen when `signupIsInitializing`?

**Review by p-jackson — COMMENTED** [review]
I like setting the flag in the middleware chain. A bit more self-documenting because you can see where the flag is set and then cleared. But I think setting the flag in the click handler is still maybe a little safer. Personally I'm only testing the one flow where we're setting the buggy behaviour, so it feels like we should only make code changes to that flow we're testing.

What if there's another flow where an exception is thrown in one of the middleware? That means the flag wouldn't get cleared. Or maybe there's some middleware that doesn't call `next()` in some contexts? The flag wouldn't get cleared in this situation either.

**Comment by danielbachhuber** [comment]
> What if there's another flow where an exception is thrown in one of the middleware? That means the flag wouldn't get cleared. Or maybe there's some middleware that doesn't call `next()` in some contexts? The flag wouldn't get cleared in this situation either.

Arg! 🥫🪱

**Comment by mpkelly** [comment]
I also took an interest in this transition due to the unnecessary `/start/user` step https://github.com/Automattic/wp-calypso/issues/68562. One thing I noticed is that there are a bunch of errors after clicking `Add new site`:

```
react_devtools_backend.js:4026 Warning: Failed prop type: The prop `canUserUseCustomerHome` is marked as required in `Home`, but its value is `null`.
    at Home (app:///./my-sites/customer-home/main.jsx:73:3)
    at eval (app:///./lib/analytics/with-tracking-tool/index.jsx:16:76)
...
Warning: Failed prop type: The prop `site` is marked as required in `Home`, but its value is `null`.
    at Home (app:///./my-sites/customer-home/main.jsx:73:3)
    at eval (app:///./lib/analytics/with-tracking-tool/index.jsx:16:76)
...
Warning: Failed prop type: The prop `siteId` is marked as required in `Home`, but its value is `null`.
    at Home (app:///./my-sites/customer-home/main.jsx:73:3)
    at eval (app:///./lib/analytics/with-tracking-tool/index.jsx:16:76)
...
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
    at SiteSetupList 
```
There seems to be something strange going on here. The component that is missing the `siteId` should only get rendered on routes that contain a `:site` part - it is using `getSelectedSiteId( state )` which is `null`. So it looks like our `state` and router are out of sync somehow. I've tested it on this branch and it still happens. I wasn't able to get to the bottom of it yet but just wanted to log it here. 

The last issue "state update on an unmounted component" is something that can be fixed easily. I will make a PR for that. The same warning is logged for the `Masterbar` component as well. I will make both of these safe.

File: client/signup/controller.js [hunk 3]
```diff
@@ -323,6 +320,17 @@ export default {
...
 
 		next();
+
+		// If the flow has siteId or siteSlug as query dependencies, we should not clear selected site id
+		if (
+			! providesDependenciesInQuery?.includes( 'siteId' ) &&
+			! providesDependenciesInQuery?.includes( 'siteSlug' ) &&
+			! isManageSiteFlow
+		) {
+			// We do this after `next()`; otherwise, components that are currently rendered and depend
```
Comment (mpkelly) [inline client/signup/controller.js:330]: Moving this block until after the route has changed stops currently rendered components that depend on `state.ui.selectedSiteId` from breaking.

Comment (mpkelly) [reply client/signup/controller.js:330]: I think we should include this eaead4affbee0ae6079246fa4cc81db23c2ea419 in this PR too. It just moves the state update to avoid errors.

Comment (mpkelly) [reply client/signup/controller.js:330]: I will fix the remaining "state update on an unmounted component" errors in a separate PR.

**Comment by danielbachhuber** [comment]
> The last issue "state update on an unmounted component" is something that can be fixed easily.

😱

**Comment by mpkelly** [comment]
> > The last issue "state update on an unmounted component" is something that can be fixed easily.
> 
> 😱

I decided against my change. Initially, I tried to clear the warnings with a component like this:

```jsx
import { Component } from 'react';

export class SafeUpdateComponent< P, S > extends Component< P, S > {
	private willUnmount = false;

	componentWillUnmount() {
		this.willUnmount = true;
	}

	setState< K extends keyof S >(
		state:
			| ( ( prevState: Readonly< S >, props: Readonly< P > ) => Pick< S, K > | S | null )
			| Pick< S, K >
			| S
			| null,
		callback?: () => void
	) {
		if ( ! this.willUnmount ) {
			super.setState( state, callback );
		}
	}
}

```
This is easy to slot into the existing code and only requires a small update to existing components to make them call `super.componentWillUnmount()`. However, after doing that on a couple of erring components, I then saw some errors coming from functional components using `useState`. Ok, I thought, I will create a `useSafeState` hook and make these components use that! But then I figured it would exist already and started googling. Eventually, I found [this](https://github.com/aldrinpvincent/useSafeState), which references this update that the [React team will remove this misleading error message](https://github.com/reactwg/react-18/discussions/82), and also do not recommend common workarounds (like I was going to implement). 

TLDR; we should live with these warnings.

**Comment by danielbachhuber** [comment]
Punting on this for now 🏈
